<a href="https://colab.research.google.com/github/wzz010624/Scigraphexp/blob/main/merged_network%2Btarget_authors_by_year.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pickle
with open('/content/drive/MyDrive/GNNdataset/coauthor_dict.pickle', 'rb') as f:
    coauthor_dict = pickle.load(f)

In [ ]:
import pickle
with open('/content/drive/MyDrive/GNNdataset/author_publication_sequence.pickle', 'rb') as f:
    author_publication_sequence = pickle.load(f)

In [ ]:
import pickle
#coauthor_dict[author_id][coauthor_id][paper_year] = [paper_id]
# 保存为二进制文件
with open('/content/drive/MyDrive/GNNdataset/5spancoauthor_networks.pickle', 'rb') as f:
    coauthor_networks = pickle.load(f)

In [ ]:
for k,v in coauthor_networks.items():
  print(k,v)

1971 Graph with 4322 nodes and 3968 edges
1972 Graph with 5024 nodes and 4600 edges
1973 Graph with 6054 nodes and 5516 edges
1974 Graph with 7745 nodes and 7391 edges
1975 Graph with 9295 nodes and 9026 edges
1976 Graph with 10820 nodes and 10903 edges
1977 Graph with 12501 nodes and 12876 edges
1978 Graph with 13680 nodes and 14555 edges
1979 Graph with 14623 nodes and 15565 edges
1980 Graph with 16011 nodes and 17491 edges
1981 Graph with 17699 nodes and 19992 edges
1982 Graph with 19892 nodes and 23013 edges
1983 Graph with 22770 nodes and 26903 edges
1984 Graph with 26337 nodes and 31726 edges
1985 Graph with 30213 nodes and 36844 edges
1986 Graph with 35453 nodes and 43363 edges
1987 Graph with 41186 nodes and 50741 edges
1988 Graph with 48840 nodes and 61232 edges
1989 Graph with 55967 nodes and 71999 edges
1990 Graph with 65868 nodes and 86682 edges
1991 Graph with 76743 nodes and 104392 edges
1992 Graph with 88243 nodes and 125173 edges
1993 Graph with 101214 nodes and 149796 

In [ ]:
import pandas as pd
import networkx as nx

# 创建一个空的 DataFrame
network_stats = pd.DataFrame(columns=['Year', 'Nodes', 'Edges', 'Clustering coefficient', 'Density'])

# 计算每个网络的统计量，并将它们添加到 DataFrame 中
for year, graph in coauthor_networks.items():
    print(year)
    # 获取网络的基本信息
    nodes = graph.number_of_nodes()
    edges = graph.number_of_edges()

    # 计算其他统计量
    clustering_coefficient = nx.average_clustering(graph)
    density = nx.density(graph)
    degree_centrality = pd.Series(nx.degree_centrality(graph))

    # 将所有统计量添加到 DataFrame 中
    network_stats = network_stats.append({
        'Year': year,
        'Nodes': nodes,
        'Edges': edges,
        'Clustering coefficient': clustering_coefficient,
        'Density': density,
    }, ignore_index=True)

# 将 DataFrame 存储到 CSV 文件中
network_stats.to_csv('network_stats.csv', index=False)

1971
1972
1973
1974
1975


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({
<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({
<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({
<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({
<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat 

1976
1977


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({
<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1978
1979


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({
<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1980
1981


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({
<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1982
1983


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1984


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1985


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1986


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1987


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1988


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1989


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1990


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1991


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1992


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1993


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1994


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1995


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1996


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1997


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1998


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


1999


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2000


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2001


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2002


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2003


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2004


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2005


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2006


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2007


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2008


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2009


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2010


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


2011


<ipython-input-12-53ef0d72287c>:20: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  network_stats = network_stats.append({


In [ ]:
status_df.to_csv('trim_network_stats.csv', index=False)

In [ ]:
def calculate_coauthors(author_id, current_year):
    unique_coauthors = set()
    for coauthor_id in coauthor_dict[author_id]:
        for year, paper_ids in coauthor_dict[author_id][coauthor_id].items():
            if year <= current_year:
                unique_coauthors.add(coauthor_id)

    total_coauthors = len(unique_coauthors)
    return total_coauthors

In [ ]:
import networkx as nx
import numpy as np

def calculate_network_stats(network):
    num_nodes = network.number_of_nodes()
    num_edges = network.number_of_edges()
    clustering_coefficient = nx.average_clustering(network)
    density = nx.density(network)

    network_stats = {'Nodes': num_nodes,
                     'Edges': num_edges,
                     'Clustering coefficient': clustering_coefficient,
                     'Density': density}
    
    return network_stats

In [ ]:
def find_target_authors(author_publication_sequence, career_age_5):
    target_authors = []
    for author_id, pub_seq in author_publication_sequence.items():
        start_year = min(pub_seq.keys())
        if career_age_5 == start_year + 4:
            end_year = max(pub_seq.keys())
            if end_year - start_year >= 14:
                start_year_15 = start_year + 14
                start_year_5 = start_year + 4
                pub_count_15 = sum([pub_seq.get(year, 0) for year in range(start_year, start_year_15 + 1)])
                pub_count_5 = sum([pub_seq.get(year, 0) for year in range(start_year, start_year_5 + 1)])
                if pub_count_15 >= 10 and pub_count_5>=3:
                    if calculate_coauthors(author_id,career_age_5)!=0:
                      target_authors.append(author_id)
    return target_authors

import networkx as nx
import gc
import pandas as pd
import networkx as nx

def extract_target_and_neighbors(network, target_nodes):
    nodes_to_keep = set(target_nodes)
    for node in target_nodes:
        neighbors = list(network.neighbors(node))
        nodes_to_keep.update(neighbors)
    
    subgraph = network.subgraph(nodes_to_keep)
    return subgraph

def merge_and_trim_networks(coauthor_networks, author_publication_sequence):
    merged_network = nx.Graph()
    target_authors_by_year = {}
    df = pd.DataFrame(columns=['Year', 'Target_Nodes', 'Nodes', 'Edges', 'Clustering coefficient', 'Density'])

    for year, network in coauthor_networks.items():
        career_age_5 = year
        target_authors = find_target_authors(author_publication_sequence, career_age_5)
        target_authors_by_year[year] = target_authors

        # 提取目标节点及其一阶邻居的子图
        subgraph = extract_target_and_neighbors(network, target_authors)

        # 计算子图的节点数、边数、聚集系数和密度
        network_stats = calculate_network_stats(subgraph)

        # 添加年份和目标节点的信息
        network_stats.update({'Year': year, 'Target_Nodes': len(target_authors)})

        # 将结果添加到DataFrame中
        df = df.append(network_stats, ignore_index=True)

        # 为子图中的节点和边添加新的唯一标识符
        new_node_ids = {node: f"{node}_{year}" for node in subgraph.nodes()}
        subgraph = nx.relabel_nodes(subgraph, new_node_ids)
        print(f"Year: {year}, Target_Nodes:  {len(target_authors)}, Nodes: {len(subgraph.nodes)}, Edges: {len(subgraph.edges)}")
        # 将子图中的节点和边添加到 merged_network 中
        merged_network.add_nodes_from(subgraph.nodes(data=True))
        merged_network.add_edges_from(subgraph.edges(data=True))

    return merged_network, target_authors_by_year, df

# 使用更新后的函数
merged_network, target_authors_by_year, status_df = merge_and_trim_networks(coauthor_networks, author_publication_sequence)

<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1971, Target_Nodes:  21, Nodes: 88, Edges: 103


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1972, Target_Nodes:  33, Nodes: 141, Edges: 181


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1973, Target_Nodes:  43, Nodes: 175, Edges: 196


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1974, Target_Nodes:  38, Nodes: 153, Edges: 186


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1975, Target_Nodes:  62, Nodes: 299, Edges: 422


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1976, Target_Nodes:  81, Nodes: 370, Edges: 519


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1977, Target_Nodes:  117, Nodes: 581, Edges: 895


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1978, Target_Nodes:  128, Nodes: 639, Edges: 1026


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1979, Target_Nodes:  112, Nodes: 495, Edges: 717


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1980, Target_Nodes:  146, Nodes: 659, Edges: 925


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1981, Target_Nodes:  172, Nodes: 812, Edges: 1145


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1982, Target_Nodes:  135, Nodes: 694, Edges: 1098


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1983, Target_Nodes:  180, Nodes: 865, Edges: 1295


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1984, Target_Nodes:  201, Nodes: 933, Edges: 1547


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1985, Target_Nodes:  256, Nodes: 1293, Edges: 2158


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1986, Target_Nodes:  331, Nodes: 1709, Edges: 2928


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1987, Target_Nodes:  365, Nodes: 1976, Edges: 3323


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1988, Target_Nodes:  402, Nodes: 2220, Edges: 3748


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1989, Target_Nodes:  498, Nodes: 2747, Edges: 4610


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1990, Target_Nodes:  584, Nodes: 3552, Edges: 6496


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1991, Target_Nodes:  731, Nodes: 4262, Edges: 7737


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1992, Target_Nodes:  891, Nodes: 5587, Edges: 10755


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1993, Target_Nodes:  1040, Nodes: 6375, Edges: 13158


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1994, Target_Nodes:  1320, Nodes: 8380, Edges: 17537


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1995, Target_Nodes:  1548, Nodes: 9894, Edges: 22269


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1996, Target_Nodes:  1722, Nodes: 11346, Edges: 25730


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1997, Target_Nodes:  2007, Nodes: 13211, Edges: 31039


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1998, Target_Nodes:  2010, Nodes: 13794, Edges: 33296


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 1999, Target_Nodes:  2355, Nodes: 16679, Edges: 40688


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2000, Target_Nodes:  2517, Nodes: 18441, Edges: 48809


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2001, Target_Nodes:  2851, Nodes: 22076, Edges: 64947


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2002, Target_Nodes:  3365, Nodes: 26409, Edges: 78676


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2003, Target_Nodes:  3765, Nodes: 30875, Edges: 103406


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2004, Target_Nodes:  4313, Nodes: 36661, Edges: 125580


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2005, Target_Nodes:  4877, Nodes: 45216, Edges: 168380


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2006, Target_Nodes:  5674, Nodes: 54829, Edges: 209058


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2007, Target_Nodes:  6175, Nodes: 62800, Edges: 245985


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2008, Target_Nodes:  7008, Nodes: 70837, Edges: 278069


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2009, Target_Nodes:  7114, Nodes: 75272, Edges: 304186


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2010, Target_Nodes:  6145, Nodes: 67481, Edges: 285269


<ipython-input-10-4182c8b53073>:51: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(network_stats, ignore_index=True)


Year: 2011, Target_Nodes:  4412, Nodes: 52249, Edges: 217967


In [ ]:
# 输出目标作者数量（所有年份的目标作者总和）
all_target_authors = set()
for target_authors in target_authors_by_year.values():
    all_target_authors.update(target_authors)
print("目标作者数量：", len(all_target_authors))

目标作者数量： 75745


In [ ]:
# 输出目标作者数量（所有年份的目标作者总和）
all_target_authors = set()
for target_authors in target_authors_by_year.values():
    all_target_authors.update(target_authors)
print("目标作者数量：", len(all_target_authors))

目标作者数量： 92623


Year: 1971, Target_Nodes:  23, Nodes: 92, Edges: 109
Year: 1972, Target_Nodes:  38, Nodes: 152, Edges: 191
Year: 1973, Target_Nodes:  51, Nodes: 194, Edges: 209
Year: 1974, Target_Nodes:  46, Nodes: 174, Edges: 209
Year: 1975, Target_Nodes:  71, Nodes: 323, Edges: 448
Year: 1976, Target_Nodes:  90, Nodes: 394, Edges: 542
Year: 1977, Target_Nodes:  138, Nodes: 623, Edges: 935
Year: 1978, Target_Nodes:  154, Nodes: 695, Edges: 1087
Year: 1979, Target_Nodes:  133, Nodes: 544, Edges: 766
Year: 1980, Target_Nodes:  183, Nodes: 755, Edges: 1040
Year: 1981, Target_Nodes:  204, Nodes: 894, Edges: 1260
Year: 1982, Target_Nodes:  166, Nodes: 766, Edges: 1166
Year: 1983, Target_Nodes:  232, Nodes: 1000, Edges: 1446
Year: 1984, Target_Nodes:  262, Nodes: 1101, Edges: 1743
Year: 1985, Target_Nodes:  329, Nodes: 1479, Edges: 2385
Year: 1986, Target_Nodes:  416, Nodes: 1972, Edges: 3353
Year: 1987, Target_Nodes:  498, Nodes: 2320, Edges: 3742
Year: 1988, Target_Nodes:  532, Nodes: 2588, Edges: 4257
Year: 1989, Target_Nodes:  648, Nodes: 3141, Edges: 5062
Year: 1990, Target_Nodes:  767, Nodes: 4065, Edges: 7286
Year: 1991, Target_Nodes:  904, Nodes: 4758, Edges: 8381
Year: 1992, Target_Nodes:  1127, Nodes: 6278, Edges: 11779
Year: 1993, Target_Nodes:  1268, Nodes: 6969, Edges: 13991
Year: 1994, Target_Nodes:  1603, Nodes: 9145, Edges: 18687
Year: 1995, Target_Nodes:  1919, Nodes: 10945, Edges: 23906
Year: 1996, Target_Nodes:  2151, Nodes: 12533, Edges: 27821
Year: 1997, Target_Nodes:  2565, Nodes: 14698, Edges: 33500
Year: 1998, Target_Nodes:  2675, Nodes: 15648, Edges: 36296
Year: 1999, Target_Nodes:  3114, Nodes: 18925, Edges: 45126
Year: 2000, Target_Nodes:  3236, Nodes: 20595, Edges: 54830
Year: 2001, Target_Nodes:  3673, Nodes: 24461, Edges: 69515
Year: 2002, Target_Nodes:  4378, Nodes: 29557, Edges: 85679
Year: 2003, Target_Nodes:  4753, Nodes: 33962, Edges: 109944
Year: 2004, Target_Nodes:  5367, Nodes: 39932, Edges: 133145
Year: 2005, Target_Nodes:  5947, Nodes: 48583, Edges: 175892
Year: 2006, Target_Nodes:  6741, Nodes: 58255, Edges: 219696
Year: 2007, Target_Nodes:  7252, Nodes: 66321, Edges: 255519
Year: 2008, Target_Nodes:  8199, Nodes: 74878, Edges: 289305
Year: 2009, Target_Nodes:  8306, Nodes: 79358, Edges: 315556
Year: 2010, Target_Nodes:  7279, Nodes: 71764, Edges: 298430
Year: 2011, Target_Nodes:  5185, Nodes: 55315, Edges: 230000

目标作者数量： 92623

In [ ]:
graph = merged_network
print(f"Nodes: {graph.number_of_nodes()}, Edges: {graph.number_of_edges()}")

Nodes: 726152, Edges: 2494234


In [ ]:
graph = merged_network
print(f"Nodes: {graph.number_of_nodes()}, Edges: {graph.number_of_edges()}")

Nodes: 673075, Edges: 2366059


In [ ]:
calculate_network_stats(graph)

{'Nodes': 673075,
 'Edges': 2366059,
 'Clustering coefficient': 0.6837628916997012,
 'Density': 1.044550138870488e-05}

Nodes: 726152, Edges: 2494234

In [ ]:
import pickle

# 将Python对象保存到文件中
with open("/content/drive/MyDrive/GNNdataset/merged_network_pub53", "wb") as f:
    pickle.dump(merged_network, f)

In [ ]:
import pickle

# 将Python对象保存到文件中
with open("/content/drive/MyDrive/GNNdataset/target_authors_by_year_pub53", "wb") as f:
    pickle.dump(target_authors_by_year, f)